WalletWasabi/Services/BackgroundTaskQueue.cs

```diff
@@ -0,0 +1,47 @@
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using WalletWasabi.Interfaces;
+
+namespace WalletWasabi.Services
+{
+	// source: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-6.0&tabs=visual-studio#queued-background-tasks
+	public class BackgroundTaskQueue : IBackgroundTaskQueue
+	{
+		private readonly Channel<Func<CancellationToken, ValueTask>> _queue;
+
+#warning TODO:
+
+		// TODO: put capacity into config
+		public BackgroundTaskQueue(int capacity = 1_000)
+		{
+			// Capacity should be set based on the expected application load and
```

Comment:
maybe default capacity should be 0 for low throughput? it should cause immediate back pressure and only need tuning to improve efficiency which should never really be the case for coinjoins.

We might want to consider running integration/functional tests with large, small and 0 bounds as well as `AllowSynchronousContinuations` which with a capacity of 0 may affect the execution order in interesting ways.

Reply:
this rubs me the wrong way. The code should be eventually consistent by default. I would actually prefer to add e.g. 5 sec delay in debug so you don't get it finished by accident before you are done. If you try to synchronize on the queue or even worse it is finished coincidently too soon.

But for testing sure synchronous finish is also interesting edge case.

Basically if you want to await don't put it to the queue in the first place the only point of the queue is so you don't have to await it and finish faster and also to serialize it to single thread so you don't overload the system with low priority jobs. and have faster response time for the clients' requests

Reply:
This is at-most-once background work that you don't await.

Reply:
the go like channel doesn't make much sense here. You just await your async task to get such behavior you don't need a queue for that. There is somewhere hidden task queue in the framework anyway

Reply:
fair enough. :+1: on the idea of adding 5 sec delay, tests should pass even with that in place